### PR TITLE
Resource Monitoring support for Generic SAI Extensions

### PR DESCRIPTION
**Pull request by svshah-intel**
Why I did it
HLD describing Critical Resource Monitoring (CRM) support for Generic SAI Extensions

| Repo  | PR title | State |
| ----------------- | ----------------- | ----------------- |
| sonic-swss | [resource monitoring support for generic SAI extension tables #2649](https://github.com/sonic-net/sonic-swss/pull/2649) | GitHub issue/pull request detail |

### CODE REVIEW

**Comment by zhangyanzhao** [comment]
reviewed in community.

File: doc/crm/Generic_SAI_Extensions_CRM.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  - object-type    : SAI_OBJECT_TYPE_GENERIC_PROGRAMMABLE
+  - attribute-id   : SAI_GENERIC_PROGRAMMABLE_ATTR_OBJECT_NAME
+
+respective data-plane implementing this API should return resource count available for the corresponding extension table matching the generic programmable object name
+
+### APIs for used counter book-keeping
+- CrmOrch::incCrmExtTableUsedCounter()
+- CrmOrch::decCrmExtTableUsedCounter()
+
+P4Orch extension table manager is responsible to call APIs when a specific entry for extension table is successfully created or removed.
```
Comment (madhupalu) [inline doc/crm/Generic_SAI_Extensions_CRM.md:48]: How is the default watermark handled in SONiC currently? Are there any high & low watermarks available with extension tables ?

File: doc/crm/Generic_SAI_Extensions_CRM.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+| 0.1  | 1/17/2023 | Intel                                                |  Initial Revision                  |
+
+## Introduction <a name="Introduction"></a>
+High Level Design for critical resource monitoring (crm) in SONiC for Generic SAI Extension tables.
+
+## Scope <a name="Scope"></a>
+The scope of the design is to publish CRM counts, for Generic SAI Extension tables, into the COUNTERS-DB. Support for existing generic crm configurations that is applicable to all resources, like polling interval, is in the scope of this document. No additional configuration specific to extension tables resource is in the scope.
+
+## Requirements <a name="Requirements"></a>
+- Support reporting of free counts
```
[the remark this answers is not in the document]
Comment (svshah-intel) [reply doc/crm/Generic_SAI_Extensions_CRM.md]: fixed

File: doc/crm/Generic_SAI_Extensions_CRM.md (hunk 1 above)
Comment (svshah-intel) [reply doc/crm/Generic_SAI_Extensions_CRM.md:48]: there is a default for low and high watermarks which is associated with every resource type

**Comment by svshah-intel** [comment]
@madhupalu could you please approve if looks good

File: doc/crm/Generic_SAI_Extensions_CRM.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+## Revision
+| Rev  |   Date    |                        Author                        | Change Description                 |
+| :--: | :-------: | :--------------------------------------------------: | :--------------------------------: |
+| 0.1  | 1/17/2023 | Intel                                                |  Initial Revision                  |
+
+## Introduction <a name="Introduction"></a>
+High Level Design for critical resource monitoring (crm) in SONiC for Generic SAI Extension tables.
+
+## Scope <a name="Scope"></a>
+The scope of the design is to publish CRM counts, for Generic SAI Extension tables, into the COUNTERS-DB. Support for existing generic crm configurations that is applicable to all resources, like polling interval, is in the scope of this document. No additional configuration specific to extension tables resource is in the scope.
```
Comment (madhupalu) [inline doc/crm/Generic_SAI_Extensions_CRM.md:19]: - It looks these SAI extensions specific to PINS project? is that true?

**Review by madhupalu — APPROVED** [review]

File: doc/crm/Generic_SAI_Extensions_CRM.md (hunk 3 above)
Comment (svshah-intel) [reply doc/crm/Generic_SAI_Extensions_CRM.md:19]: that's correct

**Comment by svshah-intel** [comment]
@volodymyrsamotiy can you please review the HLD

**Comment by reshmaintel** [comment]
@prsunny @volodymyrsamotiy Could you please complete this review for 202305. Thank you!

**Review by madhupalu — APPROVED** [review]

**Comment by reshmaintel** [comment]
@zhangyanzhao @prsunny  Could you please merge this PR, as Madhu has approved it. Thanks.

**Comment by oleksandrivantsiv** [comment]
Does this feature require CLI implementation? How the user should get the information from the DB?

**Comment by reshmaintel** [comment]
> Does this feature require CLI implementation? How the user should get the information from the DB?

This feature does not require CLI, and user can access via regular DB lookup commands

**Comment by svshah-intel** [comment]
@oleksandrivantsiv As Reshma reported, this feature was developed to be consumed by PINS deployment. There are no CLIs required. The info will be read from the DB.
Do you have other review comments? Can this be approved and merged?

**Pull request merged**